### PR TITLE
Add fix wall/table and fix wall/lepton for customizable wall fixes

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

This adds two new wall styles: wall/table and wall/lepton that use tabulation or scripting expressions to implement custom wall potentials. Both are fully integrated into the fix wall framework. This also fixes a bug in fix wall/morse.

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

The implementation of the two new wall styles follows the conventions for the corresponding bond styles.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

### CODE REVIEW

**Comment by akohlmey** [comment]
@athomps @sjplimp I found that adding Lepton support to code that computes forces can be as useful a tool for validating an implementation as using numerical differentiation, since Lepton can do symbolic differentiation and thus the forces are "exact". In this case, it was easy to validate the implementation against the existing wall potentials and also pair or bond styles, where available.
However, I ran into a strange situation with fix wall/morse where the forces differ more than what can be explained by numerics on the *second* MD step, but not on the first. Any ideas where the issue could be? Or where (else) to start looking? The code is valgrind clean. Please see the inputs added to the examples/wall folder.

**Comment by athomps** [comment]
I took a close look at this, but could not figure it out. It seems no matter what I did, the the force is *always* correct on the first invocation, but significantly wrong after that.  The new wrong force still depends on the new r, and will be correct if r is the same as first call. This suggests stale intermediate values, either in Lepton or in LAMMPS.

**Comment by athomps** [comment]
I suspect that the new wrong force depends on both the new value of r and the also the initial value of r, not on any other condtions, but I have not confirmed that.

**Comment by akohlmey** [comment]
> I took a close look at this, but could not figure it out. It seems no matter what I did, the the force is _always_ correct on the first invocation, but significantly wrong after that. The new wrong force still depends on the new r, and will be correct if r is the same as first call. This suggests stale intermediate values, either in Lepton or in LAMMPS.

I am quite confident that fix wall/table and fix wall/lepton are correct, because they a) reproduce the results for fix wall/harmonic and fix wall/lj126 (have not tried the others) and b) those also agree with the forces from bond style harmonic, and morse and pair style harmonic/cut, lj/cut and morse. It seems fix wall/morse is the odd one out. I suspect some uninitialized or not properly zeroed variable, but I cannot find it.

**Comment by akohlmey** [comment]
> I suspect that the new wrong force depends on both the new value of r and the also the initial value of r, not on any other condtions, but I have not confirmed that.

Found the bug.
The wall force computation was incorrectly dividing by "delta". The reason it didn't give wrong forces in the first step originates in the geometry that I had set up where delta had the value of 1 at the first step and thus the forces was correct. But on the second step, after the atoms had moved, the delta was != 1 and then it led to wrong forces.

File: doc/src/fix_wall.rst [hunk 1]
```diff
@@ -266,20 +360,19 @@ define the lattice spacings.
...
+manner.
+
+The *pbc* keyword can be used with a *yes* setting to allow walls to be
+specified in a periodic dimension.  See the :doc:`boundary <boundary>`
+command for options on simulation box boundaries.  The default for *pbc*
+is *no*, which means the system must be non-periodic when using a wall.
+But you may wish to use a periodic box.  E.g. to allow some particles to
+interact with the wall via the fix group-ID, and others to pass through
+it and wrap around a periodic box.  In this case you should ensure that
+the wall if sufficiently far enough away from the box boundary.  If you
```
Comment (stanmoore1) [inline doc/src/fix_wall.rst:374]: typo: if --> is

**Review by stanmoore1 — APPROVED** [review]
I approve, found a typo in the docs though

File: doc/src/fix_wall.rst (hunk 1 above)
Comment (akohlmey) [reply doc/src/fix_wall.rst:374]: This must have been there for some time... fixing it now in #3663

**Pull request merged**